content/docs/contributing/featuregates.md

```diff
@@ -0,0 +1,48 @@
+This mechanism differs from [the one that Kubernetes uses for feature gated API field implementation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) where the field will be simply set to nil if the feature gate is disabled. We chose to use webhook validation instead to make debugging easier for users who are attempting to use the feature gated field, but have forgotten to set the feature gate.
+
+
+### Implementation
+
+- Implement the new field and document that it is feature gated and in order to use it feature gates need enabling
+- Add a new [webhook feature gate](https://github.com/cert-manager/cert-manager/blob/3a055cc2f56c1c2874807af4a8f84d0a1c46ccb4/internal/webhook/feature/features.go#L25-L39) for the field
+- Update webhook validation checks for the relevant resource kind to ensure that if the feature gated field is set, but the webhook feature gate is not enabled, the resource gets rejected
+- Add a new [controller feature gate](https://github.com/cert-manager/cert-manager/blob/2417132b3cd017b5f0974006e03c2b8a540efe3f/internal/controller/feature/features.go#L26-L54) for the field
+- Ensure that any control loops that use the feature, check that the feature gate is actually enabled. (This is required to cover edge cases such if webhook runs a version of cert-manager where the feature is in GA whereas controller runs an older version where the feature is still in experimental state)
+- Ensure that the feature gate is added to cert-manager installation scripts for CI and local tests in [make](https://github.com/cert-manager/cert-manager/blob/134398e939bb2b1401697eaf589405ad469cd609/make/e2e-setup.mk#L165) and [bazel](https://github.com/cert-manager/cert-manager/blob/fd747b42b9ab4b6409b61b7946e8dc14d532e950/devel/addon/certmanager/install.sh#L26) scripts
+- Default cert-manger e2e CI tests run with all feature gates for all components enabled. There is an additional optional e2e test that runs with all feature gates disabled. You can trigger that for your PR with `/test pull-cert-manager-e2e-feature-gates-disabled` to verify that all works as expected both with and without the new feature gate.
+
+### Potential issues
+
+- A user has to remember to set two feature gates, one of the webhook one on the controller for the feature to function. Forgetting to set one of them might result in unexpected behavior
+
+- A user must remember to remove the alpha fields from their manifests when disabling a previously enabled API feature. Failing to do so might result in unexpected behavior- for example forgetting to remove an feature gated field from a `Certificate` resource might result in failed renewals at some later point when cert-manager's controller will attempt to update the `Certificate` spec, but the webhook will reject the update due to the feature gated field being set
```

Comment:
👍 your idea of only gating on CREATE and not UPDATE sounds reasonable (though I need to consider/mull over any side-effects of this..). That said, this _is_ why we advise people to not use alpha features and is exactly why they are gated. We could remove the feature *altogether* in release N+1 as well.

Reply:
Thanks I will also look into how CREATE vs UPDATE would look like

Agree that alpha could be less stable.. but I see a bunch of proposals/issues/PRs that would result in new alpha fields against API, some of which may be more popular than others.. so I think in reality we may end up with this mechanism being used quite a bit

Reply:
This can be looked into separately and probably won't happen now when a bunch of things have already been implemented using the 2 flags approach